src/EventStore.Core/Services/Storage/EpochManager/EpochManager.cs

```diff
@@ -177,101 +211,139 @@ public class EpochManager : IEpochManager {
 				if (!_writer.Write(rec, out pos))
-					throw new Exception(string.Format("Second write try failed at {0}.", epoch.EpochPosition));
+					throw new Exception($"Second write try failed at {epoch.EpochPosition}.");
 			}
-
-			Log.Debug("=== Writing E{epochNumber}@{epochPosition}:{epochId:B} (previous epoch at {lastEpochPosition}). L={leaderId:B}.",
-				epochNumber, epoch.EpochPosition, epochId, lastEpochPosition, epoch.LeaderInstanceId);
-
+			_writer.Flush();
 			_bus.Publish(new SystemMessage.EpochWritten(epoch));
 			return epoch;
 		}
 
-		public void SetLastEpoch(EpochRecord epoch) {
-			Ensure.NotNull(epoch, "epoch");
-
-			lock (_locker) {
-				if (epoch.EpochPosition > _lastEpochPosition) {
-					UpdateLastEpoch(epoch, flushWriter: false);
-					return;
-				}
-			}
-
+		public void CacheEpoch(EpochRecord epoch) {
+			InternalCacheEpoch(epoch); 
 			// Epoch record must have been already written, so we need to make sure it is where we expect it to be.
 			// If this check fails, then there is something very wrong with epochs, data corruption is possible.
 			if (!IsCorrectEpochAt(epoch.EpochPosition, epoch.EpochNumber, epoch.EpochId)) {
-				throw new Exception(string.Format("Not found epoch at {0} with epoch number: {1} and epoch ID: {2}. "
-				                                  + "SetLastEpoch FAILED! Data corruption risk!",
-					epoch.EpochPosition,
-					epoch.EpochNumber,
-					epoch.EpochId));
+				throw new Exception(
+					$"Not found epoch at {epoch.EpochPosition} with epoch number: {epoch.EpochNumber} and epoch ID: {epoch.EpochId}. " +
+					"SetLastEpoch FAILED! Data corruption risk!");
 			}
 		}
+		public void InternalCacheEpoch(EpochRecord epoch) {
+			Ensure.NotNull(epoch, "epoch");
 
-		private void UpdateLastEpoch(EpochRecord epoch, bool flushWriter) {
 			lock (_locker) {
-				_epochs[epoch.EpochNumber] = epoch;
-				_lastEpochNumber = epoch.EpochNumber;
-				_lastEpochPosition = epoch.EpochPosition;
-				_minCachedEpochNumber = Math.Max(_minCachedEpochNumber, epoch.EpochNumber - CachedEpochCount + 1);
-				_epochs.Remove(_minCachedEpochNumber - 1);
-
-				if (flushWriter)
-					_writer.Flush();
-				// Now update epoch checkpoint, so on restart we don't scan sequentially TF.
-				_checkpoint.Write(epoch.EpochPosition);
-				_checkpoint.Flush();
-
-				Log.Debug(
-					"=== Update Last Epoch E{epochNumber}@{epochPosition}:{epochId:B} (previous epoch at {lastEpochPosition}) L={leaderId:B}.",
+				
+				// if we have a first or last and either are equal it's already cached, just return
+				if ((_epochs.Last?.Value.EpochNumber ?? -1) == epoch.EpochNumber ||
+					(_epochs.First?.Value.EpochNumber ?? -1) == epoch.EpochNumber) { return; } 
+
+				//new last epoch written or received, this is the normal case
+				//if the list is empty Last will be null
+				if (_epochs.Last == null || _epochs.Last.Value.EpochNumber < epoch.EpochNumber) {
+					_epochs.AddLast(epoch);
```

Comment:
@condron _cacheSize is not being respected here - items should be removed from the head of the list if size > cacheSize

Reply:
addressed